tests/multigrid/test_hiptmair.py

```diff
@@ -0,0 +1,82 @@
+from firedrake import *
+import pytest
+
+
+def mesh_hierarchy(cell):
+    nx = 5
+    nlevels = 2
+    if cell == "tetrahedron":
+        base = UnitCubeMesh(nx, nx, nx)
+        return MeshHierarchy(base, nlevels)
+    elif cell == "hexahedron":
+        base = UnitSquareMesh(nx, nx, quadrilateral=True)
+        basemh = MeshHierarchy(base, nlevels)
+        return ExtrudedMeshHierarchy(basemh, height=1, base_layer=nx)
+
+
+def run_riesz_map(V, mat_type):
+    relax = {
+        "ksp_type": "preonly",
+        "pc_type": "jacobi",
+    }
+    coarse = {
+        "ksp_type": "preonly",
+        "pc_type": "cholesky",
+    }
+    if mat_type == "matfree":
+        coarse = {
+            "ksp_type": "preonly",
+            "pc_type": "python",
+            "pc_python_type": "firedrake.AssembledPC",
+            "assembled": coarse,
+        }
+    parameters = {
+        "mat_type": mat_type,
+        "snes_type": "ksponly",
+        "ksp_type": "cg",
+        "ksp_norm_type": "natural",
+        "pc_type": "mg",
+        "mg_coarse": coarse,
+        "mg_levels": {
+            "ksp_type": "chebyshev",
+            "ksp_maxit": 1,
+            "ksp_chebyshev_esteig": "0.75,0.25,0,1",
+            "pc_type": "python",
+            "pc_python_type": "firedrake.HiptmairPC",
+            "hiptmair_mg_levels": relax,
+            "hiptmair_mg_coarse": relax,
+        },
+    }
+
+    sobolev = V.ufl_element().sobolev_space()
+    assert sobolev in [HCurl, HDiv]
+    d = div if sobolev == HDiv else curl
+
+    u_exact = Constant((1, 2, 4))
+    f = u_exact
+
+    u = Function(V)
+    v = TestFunction(V)
+    F = inner(d(u), d(v))*dx + inner(u, v)*dx - inner(f, v)*dx
+    bcs = [DirichletBC(V, u_exact, "on_boundary")]
+
+    solve(F == 0, u, bcs=bcs, solver_parameters=parameters)
+    return norm(u_exact - u, str(sobolev))
```

Comment:
This test just tests for convergence of the discretisation. But shouldn't we test for how the solver behaved?